---
title: "Bulk_RNA_code_TKO_with_cytokines"
output: html_document
date: "06-06-2023"
---
Analysis of bulk RNA-seq data during PGCLC differentiation

Data structure consist of 10 samples.
Two time-points: EpiLCs (Day 0) and Day 6 of differentiation
Two genotypes: wild type (tg2a) and mutant (TKO)
And two conditions at day 6: with (+) or without (-) cytokines 

SET UP \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

#Libraries

```{r setup, include=FALSE}
library (DESeq2)
library(limma)
library(EnhancedVolcano)
library("ggVennDiagram")
library(org.Mm.eg.db)
library(clusterProfiler)
library(tidyverse)
library(ggplot2)
library(dplyr)
library(RColorBrewer)
library(pheatmap)
library(enrichplot)
library(dendextend)
library(ggpattern)
library(reshape)
library("factoextra")
```

#Load count matrix
In order to generate the count matrix the following steps have been performed prior to loading the data onto R:
 • Adaptor trimming with Trim galore
 • Aligment of reads to the mouse genome (mm10) wiht STAR
 • Quality filtering (Q10 filtering) with samtools 
 • And generation of count matrix with feature counts 

```{r}
setwd("~/phd/Bulk_RNA/TKO_NO_CYT/")
count_all <- read.csv("count_matrix_all.csv", row.names=1)
count_epi <- count_all[grepl("epilc", names(count_all))]
count_diff <- count_all[!grepl("epilc", names(count_all))]
colTable <- read.csv ("exp_desing.csv", sep=",", header = 1, row.names = 1)
colTable_epi <- colTable[colTable$stage=="epilc", ]
colTable_diff <- colTable[!colTable$stage=="epilc", ]
```

Analysis of all samples together -\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--
#Run deseq2 on all samples 
```{r}
colorder <- c("tg2a_cyt_D6_1"    ,   "tg2a_cyt_D6_2"   ,  "tg2a_no_cyt_epilc_1",  "tg2a_no_cyt_epilc_2"   ,  "tg2a_no_cyt_D6_1"  , "tg2a_no_cyt_D6_2","tko_epilc_2"     ,    "tko_epilc_1"         ,     "tko_D6_2"      ,      "tko_D6_1"   ,  "TKO_NO_CYT_D6_1"  ,    "TKO_NO_CYT_D6_2")

count_sample <- count_all[, colorder]
colnames(count_sample)

count_sample[is.na(count_sample)] <- 0 # set any NA values to 0
dds <- DESeqDataSetFromMatrix(countData= count_sample, 
                              colData=colTable, design= ~ line+stage+cytokines)
dds #all transcripts 25263

#selects all genes/transcripts that have more than 10 in at least one of the samples
notAllZero <- (rowSums(counts(dds))>10) 
dds <- dds[notAllZero,]
dds #sum of counts across all samples higher than 10 -> 18678

#Stablish Tg2a_no_cyt as control and run deseq2
dds$line <- relevel( dds$line, "tg2a" )
as.data.frame( colData(dds) )
dds_result <- DESeq(dds)

rld <- rlog(dds, blind=FALSE)
expression <- as.data.frame((assay(dds)))
dds <- estimateSizeFactors(dds)
counts <- counts(dds, normalized=TRUE)
```


#PCA
Generate a PCA to show similarities/differences across all samples 
```{r fig.height=5, fig.width=8}
#PCA
object <- rld
ntop <- 500
intgroup <- c("line", "stage", "cytokines")
#PCA breakdown
# calculate the variance for each gene
rv <- rowVars(assay(object))
# select the ntop genes by variance
select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]

# perform a PCA on the data in assay(x) for the selected genes
pca <- prcomp(t(assay(object)[select,]))

# the contribution to the total variance for each component
percentVar <- as.data.frame(pca$sdev^2 / sum( pca$sdev^2 ))
if (!all(intgroup %in% names(colData(object)))) {
  stop("the argument 'intgroup' should specify columns of colData(dds)")
}
intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])

# add the intgroup factors together to create a new grouping factor
group <- if (length(intgroup) > 1) {
  factor(apply( intgroup.df, 1, paste, collapse=":"))
} else {
  colData(object)[[intgroup]]
}
# assembly the data for the plot
pcaData <- data.frame(PC1=pca$x[,1], PC2=pca$x[,2], group=group, intgroup.df, name=colnames(object))

pcaData$stage <- factor(pcaData$stage,     # Reorder factor levels
                         c("2i", "epilc", "D2", "D6"))
pcaData$line <- factor(pcaData$line,     # Reorder factor levels
                        c("tg2a", "TKO", "Otx2"))

#Create a label for the x and y axis that specifies the % variance
x_axis <-   paste0("PC1: ",round(percentVar[1, ] * 100),"% variance")
y_axis <- paste0("PC2: ",round(percentVar[2, ] * 100),"% variance")

#Plot PCA plot 
pca_plot <- ggplot(pcaData, aes(x = PC1, y = PC2)) +
  geom_point(size =4, 
             aes(color = cytokines,
                    shape = stage, fill = ifelse(cytokines == "yes", line, "white")),  stroke = 1)+
  scale_shape_manual (values = c(21, 25, 22, 24 ))+
  scale_fill_manual (values= c("red2","darkgreen", "white"))+
  scale_color_manual(values = c("red2", "black") )+
  theme(text = element_text(size = 17)) + 
  xlab(x_axis) + ylab (y_axis)+
  theme_light()

#Save PCA plot
setwd("~/phd/thesis_figures/TKO_chapter/panels/D6_only")
pdf("PCA_D6_only.pdf", height=5, width=8)
pca_plot
dev.off
```





